### DISCUSSION TITLE
detecting difference between absolute and relative forms of an instruction

### DISCUSSION THREAD

**Discussion by 440bx** [zyantific/zydis#503] [Q&A] [answered]
Hello, 

I'd like to find a way to detect that the absolute form of an instruction is the same as its relative form.  For instance, instructions such as "ret", "push <register>", etc have the same absolute and relative forms since there is no address involved.

So far, I've "guessed" that if an instruction does _not_ have a memory operand or (has a memory operand but no displacement) then its absolute and relative forms will be the same.

My first question is: is the above statement always correct or am I overlooking a situation where the determination requires additional tests ?

My second question is: what is the simplest way (least amount of work, easy to understand, straightforward to code) of determining that an instruction has the same absolute and relative forms ?

Lastly and, just as something related, if there isn't a really very quick way of making the determination at this time then, it would be nice if a bit in the attributes was set aside to denote that the instruction's absolute and relative forms are the same.

thank you for your help.

**Comment by mappzor** [comment, marked as answer]
For whole instruction: [`ZYDIS_ATTRIB_IS_RELATIVE`](https://doc.zydis.re/master/html/group__instruction__attributes#ga835c24106ae43fedd65f126624d10a76)

Finding specific relative operand:
- immediate: [`imm.is_relative`](https://doc.zydis.re/master/html/structZydisDecodedOperandImm__#a65563abb14a6a0196df0ff540b9c3fa5)
- memory: check if base register is `EIP`/`RIP`

**Reply by 440bx** [reply]
Thank you Mappzor.

Somehow I missed that attribute.  Thank you  for pointing it out.  Much appreciated.
